### DISCUSSION TITLE
LQR Control

### DISCUSSION THREAD

**Discussion by Paramjit26** [bitcraze/discussions#1139] [Q&A]
I am Master Thesis Student in university of Houston, Texas. As part of my thesis I am working in crazyflie. I Want to implement LQR control for Crazyflie, I don't know how to proceed. From where can I understand Control Architecture of crazyflie? How to design your own controller in crazylie?

**Comment by ataffanel** [comment]
Controller and estimators have been made to be modular in the Crazyflie, the idea being that it should be possible to add your own and to maintain it easily (being able to update the Crazyflie firmware independently of your controller by having is out of tree).

We have a blog post that explains the architecture and development process in some details: https://www.bitcraze.io/2023/02/adding-an-estimator-or-controller/.

**Reply by Paramjit26** [reply]
thank you sir, this was helpful.

**Reply by Paramjit26** [reply]
so the Controller for Crazyflie we have to write in C language, rather than in python?

**Reply by Paramjit26** [reply]
what is app? that has been referred in the documentation?
https://www.bitcraze.io/2023/02/adding-an-estimator-or-controller/

**Reply by Paramjit26** [reply]
Sir I have one more question, if I have to build own controller for crazyflie, do I have to go through all the documentation of Crazyflie firmware? Or can we impliment our own controller using Crazyflie library?

**Reply by ataffanel** [reply]
Controllers generally have to be implemented in the Crazyflie in C. There has been experiment implementing controller running on the host via the radio, but it is adding a lot of complexity related to latency and bandwidth of the radio which is pretty much a field of study in itself. If your goal is to work on controller, it should be in C running in the Crazyflie.

You can find more information about apps in the documentation: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/. In a nutshell, the main idea of the app layer is to provide and easy to use entry-point to run your own code in the Crazyflie.

Ps. When you have unrelated questions, it is good to make a new thread. This makes it easy for others to find information later.

**Reply by Paramjit26** [reply]
Thank you sir!
